Make direct atomic inline children move for align-content on blocks.

fantasai opened the pull request:
#### 01412a0f52c64ba8fdd1c0f3885e44196c29c5d5
<pre>
Make direct atomic inline children move for align-content on blocks.
<a href="https://bugs.webkit.org/show_bug.cgi?id=267360">https://bugs.webkit.org/show_bug.cgi?id=267360</a>
<a href="https://rdar.apple.com/120802276">rdar://120802276</a>

Reviewed by Alan Baradlay.

Moves the RenderBox of atomic inlines, not just their InlineDisplay::Box.

* LayoutTests/imported/w3c/web-platform-tests/css/css-align/blocks/align-content-block-012-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-align/blocks/align-content-block-012.html: Added.
* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.cpp:
(WebCore::LayoutIntegration::LineLayout::shiftLinesBy): Move RenderBoxes associated with moved InlineDisplay::Boxes.

Canonical link: <a href="https://commits.webkit.org/273003@main">https://commits.webkit.org/273003@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/162cb5b00836e15570d68796d931f2e3d02ded68

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33834 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12614 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35780 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36454 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30668 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34888 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14985 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9760 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29730 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34318 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10638 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30163 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9284 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9388 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30183 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37766 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30709 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30501 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35498 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9522 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7465 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33386 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11305 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/29845 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7813 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10097 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10317 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.cpp:
```diff
@@ -1003,11 +1003,18 @@ void LineLayout::shiftLinesBy(LayoutUnit blockShift)
 
-    for (auto& box : m_inlineContent->displayContent().boxes) {
+    LayoutUnit dx = (isHorizontalWritingMode) ? 0_lu : blockShift;
+    LayoutUnit dy = (isHorizontalWritingMode) ? blockShift : 0_lu;
```
please use deltaX, deltaY according to WebKit coding style: https://webkit.org/code-style-guidelines/#names-full-words.

nt1m replied:
Also can we remove the redundant brackets around `isHorizontalWritingMode` ?

fantasai replied:
Done.

alanbaradlay commented on Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.cpp:
```diff
@@ -1003,11 +1003,18 @@ void LineLayout::shiftLinesBy(LayoutUnit blockShift)
-    for (auto& box : m_inlineContent->displayContent().boxes) {
+    LayoutUnit dx = (isHorizontalWritingMode) ? 0_lu : blockShift;
+    LayoutUnit dy = (isHorizontalWritingMode) ? blockShift : 0_lu;
+    for (InlineDisplay::Box& box : m_inlineContent->displayContent().boxes) {
```
(while I understanding the reason)This change is not inline with WebKit's usage of auto.

alanbaradlay replied:
(if you would like to change the way WebKit uses auto, please start a discussion at webkit-dev)

fantasai replied:
Where is WebKit's guidelines for the use of auto documented? Because it's not in https://webkit.org/code-style-guidelines/

fantasai replied:
(And yes, I can change it over, but this function is really incomprehensible without explicit typing...)

nt1m replied:
There was some prior discussion on the webkit-dev mailing list

alanbaradlay replied:
correct, it's not in there (not sure why though), I meant it more of a general direction based on prior discussions.

nt1m replied:
This was relatively recent discussion (2022-2021?), so probably no-one has had time to properly write it down.

fantasai replied:
Ok, I'll go poking around for that thread at some point.

fantasai committed:
Make direct atomic inline children move for align-content on blocks.
https://bugs.webkit.org/show_bug.cgi?id=267360
rdar://120802276

Reviewed by Alan Baradlay.

Moves the RenderBox of atomic inlines, not just their InlineDisplay::Box.

* LayoutTests/imported/w3c/web-platform-tests/css/css-align/blocks/align-content-block-012-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-align/blocks/align-content-block-012.html: Added.
* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.cpp:
(WebCore::LayoutIntegration::LineLayout::shiftLinesBy): Move RenderBoxes associated with moved InlineDisplay::Boxes.

Canonical link: https://commits.webkit.org/273003@main
